lib/web_ui/lib/src/engine/pointer_binding.dart

```diff
@@ -288,41 +288,84 @@ class _SanitizedDetails {
 class _ButtonSanitizer {
   int _pressedButtons = 0;
 
-  // Transform html.PointerEvent.buttons to Flutter's PointerEvent buttons.
+  /// Transform [html.PointerEvent.buttons] to Flutter's PointerEvent buttons.
   int _htmlButtonsToFlutterButtons(int buttons) {
     // Flutter's button definition conveniently matches that of JavaScript
     // from primary button (0x1) to forward button (0x10), which allows us to
     // avoid transforming it bit by bit.
     return buttons & _kButtonsMask;
   }
 
-  List<_SanitizedDetails> sanitizeDownEvent({@required int buttons}) {
-    final List<_SanitizedDetails> result = <_SanitizedDetails>[];
+  /// Given [html.PointerEvent.button] and [html.PointerEvent.buttons], tries to
+  /// infer the correct value for Flutter buttons.
+  int _inferDownFlutterButtons(int button, int buttons) {
+    if (buttons == 0 && button > -1) {
+      // In some cases, the browser sends `buttons:0` in a down event. In such
+      // case, we try to infer the value from `button`.
+      buttons = convertButtonToButtons(button);
+    }
+    return _htmlButtonsToFlutterButtons(buttons);
+  }
+
+  List<_SanitizedDetails> sanitizeDownEvent({
+    @required int button,
+    @required int buttons,
+  }) {
     // TODO(flutter_web): Remove this temporary fix for right click
     // on web platform once context gesture is implemented.
+
+    // If the pointer is already down, we just send a move event with the new
+    // `buttons` value.
     if (_pressedButtons != 0) {
-      _pressedButtons = 0;
-      result.add(_SanitizedDetails(
-        change: ui.PointerChange.up,
-        buttons: 0,
-      ));
+      return sanitizeMoveEvent(buttons: buttons);
     }
-    _pressedButtons = _htmlButtonsToFlutterButtons(buttons);
-    result.add(_SanitizedDetails(
-      change: ui.PointerChange.down,
-      buttons: _pressedButtons,
-    ));
-    return result;
+
+    _pressedButtons = _inferDownFlutterButtons(button, buttons);
+    return <_SanitizedDetails>[
+      _SanitizedDetails(
+        change: ui.PointerChange.down,
+        buttons: _pressedButtons,
+      )
+    ];
   }
 
   List<_SanitizedDetails> sanitizeMoveEvent({@required int buttons}) {
```

Comment:
I feel we should move the sanitizing logic in one place, either here or in pointer_converter.dart

Reply:
There are two types of sanitization going on:

1. Given a browser event and deciding which `PointerChange` value to send to flutter.
2. Synthesizing additional `PointerChange` values based on the current state.

Currently, we do `#1` in `pointer_binding.dart` and `#2` in `pointer_converter.dart`. What I like about the current structure is that `pointer_converter.dart` has no knowledge about browser events. It only operates on `PointerData`, `PointerChange`, etc.